### ISSUE TITLE
StratifiedKFold shuffle shuffles only batch order; within-stratum permutations are coordinated across classes (bug)

### ISSUE THREAD

**Issue by rowan-stein** [agyn-sandbox/scikit-learn#21]
### User request
sklearn.model_selection.StratifiedKFold either shuffling is wrong or documentation is misleading.

Documentation says shuffle shuffles each stratification, but current behavior shuffles order of batches, producing identical pairs within strata across seeds. Regression from #7823: using same random_state per stratum.

Expected:
Different shuffled splits per random_state; shuffling should be within strata, not only batch order.

### Research specification
- File: `sklearn/model_selection/_split.py`
- Class: `StratifiedKFold`
- Method: `_make_test_folds(self, X, y=None)` builds per-class KFold generators and zips their test indices.

#### Current behavior and root cause
- When `shuffle=True` and `random_state` is an integer, a fresh RNG is created inside each per-class KFold using the same seed. For classes with equal sizes, each KFold shuffles arrays of the same length with identical RNG streams, yielding identical within-stratum permutations across classes. This makes shuffling appear to only change batch/fold order and not within-stratum pairing.
- When `random_state` is a `RandomState` instance, shared state advances across classes and permutations differ. The problematic case is integer seeds.
- This behavior is consistent with changes introduced around upstream PR #7823.

#### Proposed fix
- When `shuffle=True`, derive independent per-class seeds from a base RNG seeded by `random_state`, so per-class permutations are independent while still deterministic.
- Concretely:
  - `base_rng = check_random_state(self.random_state)`
  - `class_seeds = base_rng.randint(0, np.i32(np.iinfo(np.int32).max), size=len(y_counts))`
  - Build per-class KFold with `random_state=seed` for each class.
- Alternatively, pass a single `RandomState` instance to all per-class KFold to advance state between classes; the derived-seed approach avoids iteration-order sensitivity.
- No change when `shuffle=False`.

#### Tests to add
1. `test_stratifiedkfold_shuffle_independent_per_class_permutation` — With equal class sizes and `shuffle=True`, integer seed yields different per-class permutations (`perm0 != perm1`).
2. `test_stratifiedkfold_shuffle_different_seeds_change_permutations` — Different integer seeds produce different permutations within each class.
3. `test_stratifiedkfold_no_shuffle_unchanged` — Ensure behavior is unchanged when `shuffle=False`.

#### Documentation
- Update `StratifiedKFold` docstring to clarify that shuffling is performed independently within each class when `shuffle=True`, and explain `random_state` behavior for int vs `RandomState`.

### Reproduction steps
```python
import numpy as np
from sklearn.model_selection import StratifiedKFold

y = np.array([0]*20 + [1]*20)
X = np.zeros_like(y)

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
idx0 = np.where(y == 0)[0]
idx1 = np.where(y == 1)[0]
perm0, perm1 = [], []
for _, test_idx in skf.split(X, y):
    perm0.extend([np.where(idx0 == i)[0][0] for i in test_idx if y[i] == 0])
    perm1.extend([np.where(idx1 == i)[0][0] for i in test_idx if y[i] == 1])

print(perm0)
print(perm1)
# Current behavior (bug): perm0 == perm1 for many integer seeds
```

### Observed failure
With the current branch, the concatenated within-class permutations are identical (or nearly identical), indicating coordinated permutations across classes when class sizes are equal. This contradicts documentation implying per-class shuffling.

We will implement the fix, add tests and doc updates, and open a PR targeting `scikit-learn__scikit-learn-13124`.
